File: apps/ts/packages/shared/src/types/jquants.ts
// JQuants API v2 Types based on official documentation

// Authentication Types (v2 uses API key only)
export interface JQuantsConfig {
  apiKey: string;
}

// Listed Stock Information Types (eq-master)
export interface JQuantsListedInfo {
  Date: string; // YYYY-MM-DD
  Code: string;
  CoName: string;
  CoNameEn: string;
  S17: string; // Sector17Code
  S17Nm: string; // Sector17CodeName
  S33: string; // Sector33Code
  S33Nm: string; // Sector33CodeName
  ScaleCat: string; // ScaleCategory
  Mkt: string; // MarketCode
  MktNm: string; // MarketCodeName
  Mrgn?: string; // MarginCode (Standard/Premium plan only)
  MrgnNm?: string; // MarginCodeName (Standard/Premium plan only)
}

export interface JQuantsListedInfoResponse {
  data: JQuantsListedInfo[];
  pagination_key?: string;
}

// Daily Stock Quotes Types (eq-bars-daily)
export interface JQuantsDailyQuote {
  Date: string; // YYYY-MM-DD
  Code: string;
  O: number | null; // Open
  H: number | null; // High
  L: number | null; // Low
  C: number | null; // Close
  UL: number | null; // Upper Limit
  LL: number | null; // Lower Limit
  Vo: number | null; // Volume
  Va: number | null; // Value (TurnoverValue)
  AdjFactor: number; // Adjustment Factor
  AdjO: number | null; // Adjusted Open
  AdjH: number | null; // Adjusted High
  AdjL: number | null; // Adjusted Low
  AdjC: number | null; // Adjusted Close
  AdjVo: number | null; // Adjusted Volume
  // Morning session (Premium plan)
  MO?: number | null; // Morning Open
  MH?: number | null; // Morning High
  ML?: number | null; // Morning Low
  MC?: number | null; // Morning Close
  MUL?: number | null; // Morning Upper Limit
  MLL?: number | null; // Morning Lower Limit
  MVo?: number | null; // Morning Volume
  MVa?: number | null; // Morning Value
  MAdjO?: number | null; // Morning Adjusted Open
  MAdjH?: number | null; // Morning Adjusted High
  MAdjL?: number | null; // Morning Adjusted Low
  MAdjC?: number | null; // Morning Adjusted Close
  MAdjVo?: number | null; // Morning Adjusted Volume
  // Afternoon session (Premium plan)
  AO?: number | null; // Afternoon Open
  AH?: number | null; // Afternoon High
  AL?: number | null; // Afternoon Low
  AC?: number | null; // Afternoon Close
  AUL?: number | null; // Afternoon Upper Limit
  ALL?: number | null; // Afternoon Lower Limit
  AVo?: number | null; // Afternoon Volume
  AVa?: number | null; // Afternoon Value
  AAdjO?: number | null; // Afternoon Adjusted Open
  AAdjH?: number | null; // Afternoon Adjusted High
  AAdjL?: number | null; // Afternoon Adjusted Low
  AAdjC?: number | null; // Afternoon Adjusted Close
  AAdjVo?: number | null; // Afternoon Adjusted Volume
}

export interface JQuantsDailyQuotesResponse {
  data: JQuantsDailyQuote[];
  pagination_key?: string;
}

// Weekly Margin Interest Types (mkt-margin-int)
export interface JQuantsWeeklyMarginInterest {
  Date: string; // YYYY-MM-DD
  Code: string;
  ShrtVol: number; // Short Margin Trade Volume
  LongVol: number; // Long Margin Trade Volume
  ShrtNegVol: number; // Short Negotiable Margin Trade Volume
  LongNegVol: number; // Long Negotiable Margin Trade Volume
  ShrtStdVol: number; // Short Standardized Margin Trade Volume
  LongStdVol: number; // Long Standardized Margin Trade Volume
  IssType: string; // Issue Type: "1": Credit, "2": Lending/Borrowing, "3": Other
}

export interface JQuantsWeeklyMarginInterestResponse {
  data: JQuantsWeeklyMarginInterest[];
  pagination_key?: string;
}

// Query Parameters Types
export interface JQuantsListedInfoParams {
  code?: string; // 4-5 digit stock code
  date?: string; // YYYYMMDD or YYYY-MM-DD
  pagination_key?: string;
}

export interface JQuantsDailyQuotesParams {
  code?: string;
  from?: string; // YYYYMMDD or YYYY-MM-DD
  to?: string; // YYYYMMDD or YYYY-MM-DD
  date?: string; // YYYYMMDD or YYYY-MM-DD
  pagination_key?: string;
}

export interface JQuantsWeeklyMarginInterestParams {
  code?: string;
  from?: string; // YYYYMMDD or YYYY-MM-DD
  to?: string; // YYYYMMDD or YYYY-MM-DD
  date?: string; // YYYYMMDD or YYYY-MM-DD
  pagination_key?: string;
}

// Trading Calendar Types (mkt-cal)
export interface JQuantsTradingCalendar {
  Date: string; // YYYY-MM-DD
  HolDiv: string; // Holiday Division
}

export interface JQuantsTradingCalendarResponse {
  data: JQuantsTradingCalendar[];
}

export interface JQuantsTradingCalendarParams {
  hol_div?: string; // v2 parameter name
  from?: string; // YYYYMMDD or YYYY-MM-DD
  to?: string; // YYYYMMDD or YYYY-MM-DD
}

// Index Data Types (idx-bars-daily)
export interface JQuantsIndex {
  Date: string; // YYYY-MM-DD
  Code: string;
  O: number | null; // Open
  H: number | null; // High
  L: number | null; // Low
  C: number | null; // Close
}

export interface JQuantsIndicesResponse {
  data: JQuantsIndex[];
  pagination_key?: string;
}

export interface JQuantsIndicesParams {
  code?: string;
  date?: string; // YYYYMMDD or YYYY-MM-DD
  from?: string; // YYYYMMDD or YYYY-MM-DD
  to?: string; // YYYYMMDD or YYYY-MM-DD
  pagination_key?: string;
}

// TOPIX Types (idx-bars-daily-topix)
export interface JQuantsTOPIX {
  Date: string; // YYYY-MM-DD
  O: number; // Open
  H: number; // High
  L: number; // Low
  C: number; // Close
}

export interface JQuantsTOPIXResponse {
  data: JQuantsTOPIX[];
  pagination_key?: string;
}

export interface JQuantsTOPIXParams {
  from?: string; // YYYYMMDD or YYYY-MM-DD
  to?: string; // YYYYMMDD or YYYY-MM-DD
  pagination_key?: string;
}

// Financial Statements Types (fin-summary)
export interface JQuantsStatement {
  DiscDate: string; // Disclosed Date (YYYY-MM-DD)
  DiscTime: string; // Disclosed Time (HH:MM:SS)
  Code: string; // Stock Code
  DiscNo: string; // Disclosure Number
  DocType: string; // Document Type (Q1, Q2, Q3, FY, etc.)
  CurPerType: string; // Current Period Type
  CurPerSt: string; // Current Period Start Date
  CurPerEn: string; // Current Period End Date
  CurFYSt: string; // Current Fiscal Year Start Date
  CurFYEn: string; // Current Fiscal Year End Date
  NxtFYSt: string | null; // Next Fiscal Year Start Date
  NxtFYEn: string | null; // Next Fiscal Year End Date
  // Financial Performance
  Sales: number | null; // Net Sales
  OP: number | null; // Operating Profit
  OdP: number | null; // Ordinary Profit
  NP: number | null; // Net Profit
  EPS: number | null; // Earnings Per Share
  DEPS: number | null; // Diluted EPS
  // Financial Position
  TA: number | null; // Total Assets
  Eq: number | null; // Equity
  EqAR: number | null; // Equity to Asset Ratio
  BPS: number | null; // Book Value Per Share
  // Cash Flow
  CFO: number | null; // Cash Flow from Operating Activities
  CFI: number | null; // Cash Flow from Investing Activities
  CFF: number | null; // Cash Flow from Financing Activities
  CashEq: number | null; // Cash and Equivalents
  // Dividend Information - Results
  Div1Q: number | null; // Result Dividend Per Share 1st Quarter
  Div2Q: number | null; // Result Dividend Per Share 2nd Quarter
  Div3Q: number | null; // Result Dividend Per Share 3rd Quarter
  DivFY: number | null; // Result Dividend Per Share Fiscal Year End
  DivAnn: number | null; // Result Dividend Per Share Annual
  DivUnit: number | null; // Distributions Per Unit
  DivTotalAnn: number | null; // Result Total Dividend Paid Annual
  PayoutRatioAnn: number | null; // Result Payout Ratio Annual
  // Dividend Information - Forecasts
  FDiv1Q: number | null; // Forecast Dividend Per Share 1st Quarter
  FDiv2Q: number | null; // Forecast Dividend Per Share 2nd Quarter
  FDiv3Q: number | null; // Forecast Dividend Per Share 3rd Quarter
  FDivFY: number | null; // Forecast Dividend Per Share Fiscal Year End
  FDivAnn: number | null; // Forecast Dividend Per Share Annual
  FDivUnit: number | null; // Forecast Distributions Per Unit
  FDivTotalAnn: number | null; // Forecast Total Dividend Paid Annual
  FPayoutRatioAnn: number | null; // Forecast Payout Ratio Annual
  // Next Year Dividend Forecasts
  NxFDiv1Q: number | null;
  NxFDiv2Q: number | null;
  NxFDiv3Q: number | null;
  NxFDivFY: number | null;
  NxFDivAnn: number | null;
  NxFDivUnit: number | null;
  NxFPayoutRatioAnn: number | null;
  // Performance Forecasts - 2Q
  FSales2Q: number | null;
  FOP2Q: number | null;
  FOdP2Q: number | null;
  FNP2Q: number | null;
  FEPS2Q: number | null;
  // Next Year Performance Forecasts - 2Q
  NxFSales2Q: number | null;
  NxFOP2Q: number | null;
  NxFOdP2Q: number | null;
  NxFNp2Q: number | null;
  NxFEPS2Q: number | null;
  // Performance Forecasts - FY
  FSales: number | null;
  FOP: number | null;
  FOdP: number | null;
  FNP: number | null;
  FEPS: number | null;
  // Next Year Performance Forecasts - FY
  NxFSales: number | null;
  NxFOP: number | null;
  NxFOdP: number | null;
  NxFNp: number | null;
  NxFEPS: number | null;
  // Changes and Adjustments
  MatChgSub: boolean | null; // Material Changes In Subsidiaries
  ChgByASRev: boolean | null; // Changes by Accounting Standards Revision
  ChgNoASRev: boolean | null; // Changes Not by Accounting Standards Revision
  ChgAcEst: boolean | null; // Changes In Accounting Estimates
  RetroRst: boolean | null; // Retrospective Restatement
  // Share Information
  ShOutFY: number | null; // Shares Outstanding at FY End (incl. Treasury)
  TrShFY: number | null; // Treasury Stock at FY End
  AvgSh: number | null; // Average Number of Shares
  // Non-Consolidated Data
  NCSales: number | null;
  NCOP: number | null;
  NCOdP: number | null;
  NCNP: number | null;
  NCEPS: number | null;
  NCTA: number | null;
  NCEq: number | null;
  NCEqAR: number | null;
  NCBPS: number | null;
  // Non-Consolidated Forecasts - 2Q
  FNCSales2Q: number | null;
  FNCOP2Q: number | null;
  FNCOdP2Q: number | null;
  FNCNP2Q: number | null;
  FNCEPS2Q: number | null;
  // Non-Consolidated Next Year Forecasts - 2Q
  NxFNCSales2Q: number | null;
  NxFNCOP2Q: number | null;
  NxFNCOdP2Q: number | null;
  NxFNCNP2Q: number | null;
  NxFNCEPS2Q: number | null;
  // Non-Consolidated Forecasts - FY
  FNCSales: number | null;
  FNCOP: number | null;
  FNCOdP: number | null;
  FNCNP: number | null;
  FNCEPS: number | null;
  // Non-Consolidated Next Year Forecasts - FY
  NxFNCSales: number | null;
  NxFNCOP: number | null;
  NxFNCOdP: number | null;
  NxFNCNP: number | null;
  NxFNCEPS: number | null;
}

export interface JQuantsStatementsResponse {
  data: JQuantsStatement[];
  pagination_key?: string;
}

export interface JQuantsStatementsParams {
  code?: string; // 4-5 digit stock code. Either code or date is required
  date?: string; // YYYYMMDD or YYYY-MM-DD. Either code or date is required
  pagination_key?: string;
}

// API Error Response
export interface JQuantsErrorResponse {
  message: string;
  status: number;
}

File: apps/bt/src/server/schemas/jquants.py
"""
JQuants Proxy Response Schemas

Hono hono-openapi-baseline.json と完全互換のレスポンススキーマ。
"""

from __future__ import annotations

from typing import Annotated, Any

from pydantic import BaseModel, BeforeValidator, Field


def _empty_str_to_none(v: Any) -> Any:
    """JQuantsの空文字数値をNoneに変換"""
    if v == "":
        return None
    return v


NullableFloat = Annotated[float | None, BeforeValidator(_empty_str_to_none)]


# --- Auth Status ---


class AuthStatusResponse(BaseModel):
    """JQuants API v2 認証ステータス"""

    authenticated: bool = Field(description="API キーが設定されているか")
    hasApiKey: bool = Field(description="API キーが設定されているか")


# --- Daily Quotes ---


class DailyQuoteItem(BaseModel):
    """日足クォートデータ（JQuants 生フォーマット）"""

    Date: str
    Code: str
    O: float | None = None  # noqa: E741
    H: float | None = None
    L: float | None = None
    C: float | None = None
    UL: float | None = None
    LL: float | None = None
    Vo: float | None = None
    Va: float | None = None
    AdjFactor: float
    AdjO: float | None = None
    AdjH: float | None = None
    AdjL: float | None = None
    AdjC: float | None = None
    AdjVo: float | None = None


class DailyQuotesResponse(BaseModel):
    """日足クォートレスポンス"""

    data: list[DailyQuoteItem]
    pagination_key: str | None = None


# --- Indices ---


class ApiIndex(BaseModel):
    """指数データポイント"""

    date: str = Field(description="日付 (YYYY-MM-DD)")
    code: str | None = None
    open: float = Field(gt=0)
    high: float = Field(gt=0)
    low: float = Field(gt=0)
    close: float = Field(gt=0)


class ApiIndicesResponse(BaseModel):
    """指数データレスポンス"""

    indices: list[ApiIndex]
    lastUpdated: str = Field(description="最終更新日時 (ISO 8601)")


# --- Listed Info ---


class ApiListedInfo(BaseModel):
    """上場銘柄情報"""

    code: str = Field(min_length=4, max_length=4)
    companyName: str
    companyNameEnglish: str | None = None
    marketCode: str | None = None
    marketCodeName: str | None = None
    sector17Code: str | None = None
    sector17CodeName: str | None = None
    sector33Code: str | None = None
    sector33CodeName: str | None = None
    scaleCategory: str | None = None
    date: str | None = None


class ApiListedInfoResponse(BaseModel):
    """上場銘柄情報レスポンス"""

    info: list[ApiListedInfo]
    lastUpdated: str = Field(description="最終更新日時 (ISO 8601)")


# --- Margin Interest ---


class ApiMarginInterest(BaseModel):
    """週次信用取引データポイント"""

    date: str = Field(description="日付 (YYYY-MM-DD)")
    code: str = Field(min_length=4, max_length=4)
    shortMarginTradeVolume: float = Field(ge=0)
    longMarginTradeVolume: float = Field(ge=0)
    shortMarginOutstandingBalance: float | None = Field(default=None, ge=0)
    longMarginOutstandingBalance: float | None = Field(default=None, ge=0)


class ApiMarginInterestResponse(BaseModel):
    """週次信用取引データレスポンス"""

    marginInterest: list[ApiMarginInterest]
    symbol: str = Field(min_length=4, max_length=4)
    lastUpdated: str = Field(description="最終更新日時 (ISO 8601)")


# --- Statements (EPS subset) ---


class StatementItem(BaseModel):
    """財務諸表データ（EPS サブセット）"""

    DiscDate: str
    Code: str
    CurPerType: str
    CurPerSt: str
    CurPerEn: str
    EPS: NullableFloat = None
    FEPS: NullableFloat = None
    NxFEPS: NullableFloat = None
    NCEPS: NullableFloat = None
    FNCEPS: NullableFloat = None
    NxFNCEPS: NullableFloat = None


class StatementsResponse(BaseModel):
    """財務諸表レスポンス（EPS サブセット）"""

    data: list[StatementItem]
    pagination_key: str | None = None


# --- Statements Raw (complete) ---


class RawStatementItem(BaseModel):
    """財務諸表データ（完全版）"""

    # Identification
    DiscDate: str
    Code: str
    DocType: str | None = None
    # Period Information
    CurPerType: str
    CurPerSt: str
    CurPerEn: str
    CurFYSt: str | None = None
    CurFYEn: str | None = None
    NxtFYSt: str | None = None
    NxtFYEn: str | None = None
    # Financial Performance (Consolidated)
    Sales: NullableFloat = None
    OP: NullableFloat = None
    OdP: NullableFloat = None
    NP: NullableFloat = None
    EPS: NullableFloat = None
    DEPS: NullableFloat = None
    # Financial Position (Consolidated)
    TA: NullableFloat = None
    Eq: NullableFloat = None
    EqAR: NullableFloat = None
    BPS: NullableFloat = None
    # Cash Flow
    CFO: NullableFloat = None
    CFI: NullableFloat = None
    CFF: NullableFloat = None
    CashEq: NullableFloat = None
    # Share Information
    ShOutFY: NullableFloat = None
    TrShFY: NullableFloat = None
    AvgSh: NullableFloat = None
    # Forecast EPS
    FEPS: NullableFloat = None
    NxFEPS: NullableFloat = None
    # Dividend
    DivAnn: NullableFloat = None
    # Non-Consolidated Financial Performance
    NCSales: NullableFloat = None
    NCOP: NullableFloat = None
    NCOdP: NullableFloat = None
    NCNP: NullableFloat = None
    NCEPS: NullableFloat = None
    # Non-Consolidated Financial Position
    NCTA: NullableFloat = None
    NCEq: NullableFloat = None
    NCEqAR: NullableFloat = None
    NCBPS: NullableFloat = None
    # Non-Consolidated Forecast EPS
    FNCEPS: NullableFloat = None
    NxFNCEPS: NullableFloat = None


class RawStatementsResponse(BaseModel):
    """財務諸表レスポンス（完全版）"""

    data: list[RawStatementItem]
    pagination_key: str | None = None


# --- TOPIX Raw ---


class TopixRawItem(BaseModel):
    """TOPIX 生データポイント"""

    Date: str
    Open: float | None = None
    High: float | None = None
    Low: float | None = None
    Close: float | None = None


class TopixRawResponse(BaseModel):
    """TOPIX 生データレスポンス"""

    topix: list[TopixRawItem]
